### ISSUE TITLE
Redirection conflict with .htaccess when dealing with HTTPS

### ISSUE THREAD

**Issue by NicoPL** [polylang/polylang#150]
## Prerequisites

- [x] This is not a usage question (Those should be directed to the [community supported forum](https://wordpress.org/support/plugin/polylang), unless this is a question about Polylang Pro in which case you should [use the helpdesk](https://polylang.pro/support/)).
- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x] The issue still exists against the latest `master` branch of Polylang and the latest WordPress version.
- [x] This bug happens with only Polylang plugin active
- [x] This bug happens with a default WordPress theme active
- [x] I can reproduce this bug consistently

## Steps to reproduce the issue

1. Install Polylang
2. Have a .htaccess to redirect HTTPS
`RewriteEngine On
RewriteCond %{HTTPS} =off
RewriteRule ^(.*)$ https://mydomain.com/$1 [L,R=301]`
3. Once on your WP website (For example https://mydomain.com/wordpress/whatever/whatever/) try to remove the "s" of "https" and see what happen.

## Expected behavior and actual behavior

When I follow those steps, I see that the redirection towards https does not work and it the URL stays "http" instead of becoming "https"

I was expecting the "http" become "https".

I confirm that this issue appears only when Polylang is active, and whatever the Permalinks settings. (Either having /en/ in the URL or using "from content").

My investigations showed that there is also another issue to deal with redirections.
When Polylang is active, and trying to access a URI that does not exists (for example mydomain.com/mywordpress/en/blabla ), it redirects to mydomain.com/en instead of mydomain/mywordpress/en

Thanks in advance for your help fixing this issue, and I'll be happy to help for further investigations on this issue.

**Comment by Chouby** [comment]
I tested your rewrite rule in .htaccess on my local install and it works as expected. So I suppose that there is something wrong in your install.

Moreover, the redirect in .htaccess are done directly by Apache, much before WordPress is loaded. So I don't see how Polylang could interfer in this process.

**Comment by NicoPL** [comment]
Thank you for testing.
Then I really don't what happens. Because the issue occurs only once
Polylang is enable.
I have a domain, and two websites in it.
- mydomain.com/website
- mydomain.com/wordpress
My .htaccess is in my public_html, so the root folder of my domain www.
I confirm that I have no HTTPS redirection issue whatever the url is after
mydomain.com/website. I confirm also that I have no issue in
mydomain.com/wordpress ONLY when Polylang is unactive.

And I confirm that this issue occurs also on the wordpress one of my friend
with Polylang.

Polylang rewrites the URL from mydomain.com/wordpress to
mydomain.com/wordpress/en/ or mydomain.com/wordpress/fr or whatever
depending on the language selected. I assume that a php function does that.
It seems occurring after the .htaccess rules.
For instance, I checked that having an index.php file this
"header("Location: http://mydomain.com/wordpress/");" will redirect to http
even if I have the https redirection in my .htaccess.

**Comment by Chouby** [comment]
> For instance, I checked that having an index.php file this
"header("Location: http://mydomain.com/wordpress/");" will redirect to http
even if I have the https redirection in my .htaccess.

Yes. In that case you would have 2 redirects which possibly do 2 opposite things.

Polylang checks `is_ssl()` before it decides wether to redirect to http or https. I suggest that you install a plugin like https://wordpress.org/plugins/really-simple-ssl/. It will manage correctly the SSL on your WP install. You can keep the redirect in .htaccess as this plugin manages the redirect in PHP and yours in .htaccess is way much faster :)

**Comment by NicoPL** [comment]
Thank you for the advice about RSSSL, I'll look at it.
I don't use any php function to redirect (the one in my previous message
was only for the example :)

Then don't you think it could be this "is_ssl()" function that could brings
this issue ?

For example:
- I go to the address https://mydomain.com/wordpress/en , everything is fine
- I decide to remove the "s" in the URI. Then writing
http://mydomain.com/wordpress/en .
- The result I still get is that the http://mydomain.com/wordpress/en is
reached and is not redirected to https://mydomain.com/wordpress/en

I understand that you said that Apache redirects before php scripts are
executed so "is_ssl()" should give True and not False in my case.

**Comment by Chouby** [comment]
You are right. An issue with `is_ssl()` is not excluded. I had a customer who experienced another issue (which I fixed since) which was due to `is_ssl()` being unreliable on some servers. The function is provided by WordPress. Several bug reports have been filed on WordPress trac, all closed. Have a look at https://core.trac.wordpress.org/ticket/40710#comment:1
